### ISSUE TITLE
Track windows opened through exec for command criteria

### ISSUE THREAD

**Issue by Airblader** [i3/i3#3895]
## I'm submitting a…

<pre>
[ ] Bug
[x] Feature Request
[ ] Documentation Request
[ ] Other (Please describe in detail)
</pre>

## Current Behavior
There is no generic way of running commands for a window opened through `exec` unless you have a-priori knowledge of the specific application (class, instance, …).

This is a very frequent request on our support reddit, so I think we can think about solutions to improve this in core i3.

## Desired Behavior
We already support startup notifications, so we could do something like

```
exec --trace-startup "foo" "/pathh/to/foo"
```

In combination with some new command criterion users could have something like

```
for_window [startup_trace=foo] floating enable
```

in their config and this would now be independent of the actual application. Note that this only works for applications implementing startup notifications and thus we should also throw an error if the flag is used in combination with `--no-startup-id` (of course this still does not guarantee that it will work, so we need to clearly document the expectation here).

CC @stapelberg @orestisfl What do you think?

Output of `i3 --moreversion 2>&-`:
<pre>
i3 version: 4.17
</pre>

**Comment by orestisfl** [comment]
This doesn't sound too complicated so I am in favor. It would be even simpler to place a user-provided mark on the window (#1682) to reduce new syntax additions:

`exec --trace-startup "foo" "/pathh/to/foo"`  
`for_window [con_mark=foo] floating enable`

**Comment by Airblader** [comment]
Yeah I was considering "hard-wiring" it to use marks as well, the main issue here is that multiple windows opening would steal the mark from each other.

(My "trace-startup" naming is definitely not a final suggestion but just a working title, though)

**Comment by stapelberg** [comment]
Sounds good in principle.

Using unique identifiers (e.g. returned by the exec command via IPC) would increase robustness, but slightly decrease convenience. Maybe if the identifier is empty, we’ll return one, so that users can chose which approach they prefer?

**Comment by Airblader** [comment]
Sure, sounds good with me.

I think the only part left in this case would be naming of the `exec` flag and the command criterion. Marking as accepted, though.

**Comment by orestisfl** [comment]
> Maybe if the identifier is empty, we’ll return one

This sounds a bit complicated for the parser, we already have a string in the end of the `exec` command, e.g. consider `exec --flag myexec` vs `exec --flag myexec myexec`.

**Comment by stapelberg** [comment]
We could also use a special identifier, similar to `__focused__`

**Comment by paoloose** [comment]
Hello!,

> There is no generic way of running commands for a window opened through exec unless you have a-priori knowledge of the specific application (class, instance, …).

What is the status of this? I think it could be a very powerful feature for working with multiple instances of the same application, like Chrome profiles. Also with different instances of different applications but with the same category or semantically "grouped".

Of course, many users will probably use this feature unnecessarily when the window's title and class are sufficient. But when they aren't this is very handy.

The [stapelberg's comment](https://github.com/i3/i3/issues/1682#issuecomment-98503046) in #1682 is very related:

> I think if there are applications which neither set their class, nor instance, nor title to something useful for matching, that problem should be fixed upstream (i.e. in the app) since it affects all window managers and plenty of features.

Although this is true, I believe most of the users who encounter this "problem" expect this to be a feature of i3 and not the applications themselves.

So at my opinion, this remains as a very specific feature for most users, but powerful for others. "Marks" are useful but a startup tracing is missing. Reworking the "marks" concept could be interesting to solve this.

By the way, thank you guys for such an amazing tool!

**Comment by Airblader** [comment]
The specification demands that windows set a useful WM_CLASS. The reality is that not everyone follows this, but in general I would very much assume that any such client wouldn't be implementing startup notifications either, disqualifying them from this feature anyway.

By the way, we should make sure we document this really well for users. Most users don't know what "the app must support startup notifications" means, so we need to be careful about how we document this feature. I wonder if we can have automatical guidance when this feature is used for apps that don't support SN, but I can't think of a good way.
